File: frontend/src/services/tag.service.ts
import api from './api';
import { Tag } from '../types/tag';

const BASE_URL = '/tags';

export const searchTags = async (query: string) => {
  try {
    const response = await api.get(`${BASE_URL}/search?query=${encodeURIComponent(query)}`);
    return response.data as Tag[];
  } catch (error) {
    console.error('Error in searchTags:', error);
    throw error;
  }
};

export const createTag = async (name: string) => {
  try {
    const response = await api.post(BASE_URL, { name });
    return response.data as Tag;
  } catch (error) {
    console.error('Error in createTag:', error);
    throw error;
  }
};

const tagService = {
  searchTags,
  createTag,
};

export default tagService;

File: src/main/java/com/tripzin/eleganttex/service/TagService.java
package com.tripzin.eleganttex.service;

import com.tripzin.eleganttex.dto.response.TagDTO;
import com.tripzin.eleganttex.entity.Tag;
import com.tripzin.eleganttex.exception.BadRequestException;
import com.tripzin.eleganttex.repository.TagRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class TagService {
    
    private final TagRepository tagRepository;
    
    public List<TagDTO> searchTags(String query) {
        return tagRepository.findByNameContainingIgnoreCase(query).stream()
                .map(TagDTO::fromEntity)
                .collect(Collectors.toList());
    }
    
    @Transactional
    public TagDTO createTag(String name) {
        if (tagRepository.existsByName(name)) {
            throw new BadRequestException("Tag with this name already exists");
        }
        
        Tag tag = new Tag();
        tag.setName(name);
        Tag savedTag = tagRepository.save(tag);
        
        return TagDTO.fromEntity(savedTag);
    }
    
    @Transactional
    public Tag getOrCreateTag(String name) {
        return tagRepository.findByName(name)
                .orElseGet(() -> {
                    Tag newTag = new Tag();
                    newTag.setName(name);
                    return tagRepository.save(newTag);
                });
    }
}
